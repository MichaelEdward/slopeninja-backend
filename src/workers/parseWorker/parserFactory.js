import cheerio from 'cheerio';
import { AllHtmlEntities } from 'html-entities';

const htmlEntities = new AllHtmlEntities();

export const removeBackSlashes = (text) => {
  const decoded = text ? text.toString().replace(/\\/g, '') : null;
  return decoded;
}

export const xyz = (text) => {
  const decoded = text ? text.toString().replace(/\u0026amp;/gu, '__JAKE__') : null;
  return decoded;
}

export const decodeEntities = (text) => {
  // best way I found to normalize unicode in input
  const unicodeInput = JSON.stringify(JSON.parse(text.toString()));
  return htmlEntities.decode(unicodeInput);
}

export const createHtmlParser = (key, parser, preParser = t => t) => async (html) => {
  const $ = cheerio.load(preParser(html), { decodeEntities: true });

  const keyData = await parser($);

  return {
    [key]: keyData
  };
}

export const createJSONParser = (key, parser, preParser = t => t) => async (data = "{}") => {
  let json = JSON.parse(preParser(data));

  let d;
  if (json) {
    d = json;
  }

  const keyData = await parser(d);

  return {
    [key]: keyData
  };
}
