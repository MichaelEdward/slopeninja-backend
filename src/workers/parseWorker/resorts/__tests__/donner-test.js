import fs from 'fs';
import {
  parseDonnerSnow,
  parseDonnerLifts,
  parseDonnerTrails,
  parseDonnerLiftList,
  parseDonnerTrailList,
} from '../donner';
import { createHtmlParser } from '../../parserFactory';

test('fetches Donner snow data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/__fixtures__/donner-weather.html.input`);

  const resortData = await createHtmlParser('snow', parseDonnerSnow)(htmlText);
  expect(resortData).toEqual({
    snow: {
      status: null,
      weatherIcon: null,
      temprature: null,
      baseCondition: null,
      newSnow: null,
      snowDepthBase: null,
      snowDepthSummit: null,
    },
  });
})

test('fetches all null for nonexisting snow values', async () => {
  const resortData = await createHtmlParser('snow', parseDonnerSnow)('<html></html>');
  expect(resortData).toEqual({
    snow: {
      status: null,
      weatherIcon: null,
      temprature: null,
      baseCondition: null,
      newSnow: null,
      snowDepthBase: null,
      snowDepthSummit: null,
    }
  });
});

test('fetches Donner lifts data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/__fixtures__/donner-weather.html.input`);
  const resortData = await createHtmlParser('lifts', parseDonnerLifts)(htmlText);
  expect(resortData).toEqual({
    lifts: {
      total: null,
      open: null,
    }
  });
});

test('fetches all null for nonexisting lift values', async () => {
  const resortData = await createHtmlParser('lifts', parseDonnerLifts)('<html></html>');
  expect(resortData).toEqual({
    lifts: {
      total: null,
      open: null,
    }
  });
});

test('fetches Donner trails data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/__fixtures__/donner-weather.html.input`);
  const resortData = await createHtmlParser('trails', parseDonnerTrails)(htmlText);
  expect(resortData).toEqual({
    trails: {
      total: null,
      open: null,
    }
  });
});

test('fetches all null for nonexisting trails values', async () => {
  const resortData = await createHtmlParser('trails', parseDonnerTrails)('<html></html>');
  expect(resortData).toEqual({
    trails: {
      total: null,
      open: null,
    }
  });
});

test('fetches Donner lift list correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/__fixtures__/donner-weather.html.input`);
  const resortData = await createHtmlParser('liftList', parseDonnerLiftList)(htmlText);
  expect(resortData).toMatchSnapshot();
});

test('fetches all null for nonexisting lift list values', async () => {
  const resortData = await createHtmlParser('liftList', parseDonnerLiftList)('<html></html>');
  expect(resortData).toMatchObject({ liftList: [] });
});

test('fetches Donner trail list correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/__fixtures__/donner-weather.html.input`);
  const resortData = await createHtmlParser('trailList', parseDonnerTrailList)(htmlText);
  expect(resortData).toMatchObject({ trailList: [] });
});

test('fetches all null for nonexisting lift list values', async () => {
  const resortData = await createHtmlParser('trailList', parseDonnerTrailList)('<html></html>');
  expect(resortData).toMatchObject({ trailList: [] });
});
