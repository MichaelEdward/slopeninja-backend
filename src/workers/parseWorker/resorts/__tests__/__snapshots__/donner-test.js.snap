// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches Donner lift list correctly 1`] = `
Object {
  "liftList": Array [
    Object {
      "category": null,
      "name": "Chair 1",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Chair 2",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Chair 3",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Chair 4",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Chair 5",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Chair 6",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Ski Moving Carpet",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Tubing Hill",
      "status": "closed",
    },
  ],
}
`;
